pkg/connect/gateway.go

```diff
@@ -56,16 +75,28 @@ func (c *connectGatewaySvc) Handler() http.Handler {
 			return
 		}
 
+		// Do not accept new connections if the gateway is draining
+		if c.isDraining {
+			c.closeDraining(ws)
+			return
+		}
+
 		ch := &connectionHandler{
 			svc:        c,
 			log:        c.logger,
 			ws:         ws,
 			updateLock: sync.Mutex{},
 		}
 
+		c.connectionSema.Add(1)
 		defer func() {
-			c.logger.Debug("Closing WebSocket connection")
+			defer c.connectionSema.Done()
```

Comment:
why `defer` within a `defer` function?

Reply:
good one, can just always call done
```suggestion
			c.connectionSema.Done()
```

Reply:
Left a comment to describe this behavior as expected:

We always want to update the semaphore and a couple lines down, we branch out with a condition. This can be done for two cases, but it's easy to forget to update the semaphore, thus preventing a graceful shutdown. If we update the semaphore too early, we will potentially shut down without a proper reason, which is also bad.